### PR TITLE
feat(HACBS-2461): add pipelineRef struct to reference a pipeline

### PR DESCRIPTION
**Pull request by mmalina**
This will be used in release-service (and possibly others in the future) to reference a pipeline in a ReleaseStrategy CR. This definition will then be converted and used in PipelineRuns created by the release service.

### CODE REVIEW

File: tekton/pipeline_ref.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Value string `json:"value"`
+}
+
+func (pr PipelineRef) ToTektonPipelineRef() tektonv1.PipelineRef {
+	tektonParams := tektonv1.Params{}
+	for _, p := range pr.ResolverRef.Params {
+		tektonParams = append(tektonParams, tektonv1.Param{
+			Name: p.Name,
+			Value: tektonv1.ParamValue{
+				Type:      "string",
```
Comment (davidmogar) [inline tekton/pipeline_ref.go]: Use `ParamTypeString`.

File: tekton/pipeline_ref.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		})
+	}
+	tektonPipelineRef := tektonv1.PipelineRef{
+		Name: pr.Name,
+		ResolverRef: tektonv1.ResolverRef{
+			Resolver: tektonv1.ResolverName(pr.ResolverRef.Resolver),
+			Params:   tektonParams,
+		},
+	}
+	return tektonPipelineRef
```
Comment (davidmogar) [inline tekton/pipeline_ref.go:35]: Add an empty line before this one please.

File: tekton/pipeline_ref.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	for _, p := range pr.ResolverRef.Params {
+		tektonParams = append(tektonParams, tektonv1.Param{
+			Name: p.Name,
+			Value: tektonv1.ParamValue{
+				Type:      "string",
+				StringVal: p.Value,
+			},
+		})
+	}
+	tektonPipelineRef := tektonv1.PipelineRef{
```
Comment (davidmogar) [inline tekton/pipeline_ref.go:28]: Add an empty line before this one please.

File: tekton/pipeline_ref.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import tektonv1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+
+type PipelineRef struct {
+	Name        string      `json:"name,omitempty"`
+	ResolverRef ResolverRef `json:"resolverRef"`
+}
+
+type ResolverRef struct {
+	Resolver string  `json:"resolver"`
```
Comment (davidmogar) [inline tekton/pipeline_ref.go:6]: I said that we should limit the possible values, but let's avoid this to support custom resolvers.

File: tekton/pipeline_ref.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package tekton
+
+import tektonv1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+
+type PipelineRef struct {
+	Name        string      `json:"name,omitempty"`
+	ResolverRef ResolverRef `json:"resolverRef"`
```
Comment (davidmogar) [inline tekton/pipeline_ref.go]: This should be optional as well if we support the same definition Tekton supports. Another option and maybe I like it would be to remove this struct and rename ResolverRef to PipelineRef.

In fact, the RP/RPA wil have probably something like:
```
Spec:
    Applications
    PipelineRef
        Resolver
        Params
```

File: tekton/pipeline_ref.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+type Param struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+func (pr PipelineRef) ToTektonPipelineRef() tektonv1.PipelineRef {
+	tektonParams := tektonv1.Params{}
+	for _, p := range pr.ResolverRef.Params {
```
Comment (davidmogar) [inline tekton/pipeline_ref.go]:
```suggestion
	for _, param := range pr.ResolverRef.Params {
```
Also, empty line before this one please.

File: tekton/pipeline_ref.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	Params   []Param `json:"params"`
+}
+
+type Param struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+func (pr PipelineRef) ToTektonPipelineRef() tektonv1.PipelineRef {
+	tektonParams := tektonv1.Params{}
```
Comment (davidmogar) [inline tekton/pipeline_ref.go]: You can call it just `params`. Not used in this context and doesn't seem confusing to me.

File: tekton/pipeline_ref.go (hunk 4 above)
Comment (mmalina) [reply tekton/pipeline_ref.go:6]: ok. we do intend to restrict what pipelines can be used, right? but I guess that validation will go to the controller.

File: tekton/pipeline_ref.go (hunk 5 above)
Comment (mmalina) [reply tekton/pipeline_ref.go]: Yep, that's what PipelineRun.spec has as well and I thought that's how I thought the types are set up, but you correctly pointed out that the types don't match what's in the PipelineRun yaml. Just spelling it out here. 

I did as you suggested. Please take a look.

File: tekton/pipeline_ref.go (hunk 1 above)
Comment (mmalina) [reply tekton/pipeline_ref.go]: Done.

File: tekton/pipeline_ref.go (hunk 2 above)
Comment (mmalina) [reply tekton/pipeline_ref.go:35]: Fixed.

File: tekton/pipeline_ref.go (hunk 3 above)
Comment (mmalina) [reply tekton/pipeline_ref.go:28]: Done.

File: tekton/pipeline_ref.go (hunk 7 above)
Comment (mmalina) [reply tekton/pipeline_ref.go]: Done.

File: tekton/pipeline_ref.go (hunk 6 above)
Comment (mmalina) [reply tekton/pipeline_ref.go]: Done.

File: tekton/pipeline_ref.go (hunk 4 above)
Comment (davidmogar) [reply tekton/pipeline_ref.go:6]: I didn't quite get this comment, sorry. Could you rephrase it?

File: tekton/pipeline_ref.go (hunk 5 above)
Comment (mmalina) [reply tekton/pipeline_ref.go]: If you look at [PipelineRun docs](https://tekton.dev/docs/pipelines/pipelineruns/#tekton-bundles), you have something like this:
```yaml
spec:
  pipelineRef:
    resolver: bundles
    params:
    - name: bundle
      value: docker.io/myrepo/mycatalog:v1.0
```

But in the types definition, there is one extra level:
```go
type ResolverRef struct {
	Resolver ResolverName `json:"resolver,omitempty"`
	Params Params `json:"params,omitempty"`
}

type PipelineRef struct {
	Name string `json:"name,omitempty"`
	APIVersion string `json:"apiVersion,omitempty"`
	ResolverRef `json:",omitempty"`
}
```

I was just saying that originally I didn't realize this and assumed that the types are defined the way I would expect from looking at the docs (no middle layer). And that's the way we want to define it in our copy of the structure. But in any case, I did as you suggested.

**Comment by mmalina** [comment]
Since this will really only be used in the release-service, we decided not to add this here, but directly to the release-service repo: https://github.com/redhat-appstudio/release-service/pull/252
